만국박람회 [Step3] 미르, 쥬봉이

jyubong opened the pull request:
@delmaSong 
안녕하세요 델마 step3 PR입니다👐🏻
이번에도 잘 부탁드립니다! 

---
### 실행화면
|첫번째 화면|두번째-세번째화면|
|-|-|
|<img src="https://velog.velcdn.com/images/nane030/post/da15744b-a23d-4d6f-a487-61401b4c230d/image.gif" height="500">|<img src="https://velog.velcdn.com/images/nane030/post/b30f0638-aa0e-4ad2-a58f-e9b32f199887/image.gif" height="500">|

---
### 고민이 되었던 점
1. 첫 화면 세로로 보이게 구현
- `첫 화면은 세로로만 볼 수 있도록 합니다`라는 step3 과정을 통해 전체화면이 아닌 첫화면에 세로만 보이게 제약을 두어 구현하려고 했습니다.

    - `AppDelegate` 클래스에서 뷰컨트롤러에서 사용할 인터페이스 방향을 지정해주는 메서드인 [supportedInterfaceOrientationsFor](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623107-application) 를 이용하여 `ExpositionHomeViewController`에 `viewWillAppear`에는 `AppDelegate`의 프로퍼티인`sholdSupportAllOrientation`를 화면에 띄워 false를 주었고, `viewWillDisappear`에는 `sholdSupportAllOrientation`를 화면에 띄워 true를 주어 첫번째 화면에 세로로 설정해주었습니다.

~~~swift
class AppDelegate: UIResponder, UIApplicationDelegate {
    
    var sholdSupportAllOrientation = true
    
    func application(_ application: UIApplication, supportedInterfaceOrientationsFor window: UIWindow?) -> UIInterfaceOrientationMask {
        return sholdSupportAllOrientation ? .all : .portrait
    }
}
~~~
|첫번째 화면|두번째 화면|
|-|-|
|<img src ="https://hackmd.io/_uploads/r1OjHsuXa.png" width="500">|<img src="https://github.com/yagom-academy/ios-exposition-universelle/assets/126065608/615ff818-c83e-431d-9f64-cf3079eedce4" height="500">|

2. 오토레이아웃 설정
- 오토레이아웃을 최대한 `스토리보드`로 설정해주었습니다.
- **scroll view** : edge부분까지 scroll이 되어야하기때문에 superview를 기준으로 하여 0으로 설정하였습니다.
- **stack view**
    - scroll이 가로로 되어야되므로 width를 scroll view frame guide와 동일하게 설정하였습니다.
    - top은 content frame에 맞추어 navigation bar 아래에 맞춰지도록 설정하였습니다. 
    - 그 외에는 super view에 맞춰주었습니다.
- **stavk view padding** : stack view 내부 요소들도 top, botton, leadig, trailing과 간격을 주기 위해 `stack view layout margings - language directional`으로 padding을 주었습니다.
- **image view** : cell image view 외에는 스토리보드로 설정하였습니다.
    - 첫화면 포스터는 고유의 크기를 유지했습니다.
    - 버튼 옆의 image는 높이와 너비를 제한했습니다.
    - detail view에서는 높이와 너비를 제한하고 aspect Fit으로 설정한 크기내에서 원래의 비율을 유지하도록 설정하였습니다.
    - cell image view는 cell configuaration을 사용하여 60*60 안에서 기본 비율을 유지하도록 설정했습니다. 

        ```swift
        let imageWidth = 60
        let imageSize = CGSize(width: imageWidth, height: imageWidth)
        content.image = UIImage(named: item.imageName)
        content.imageProperties.maximumSize = imageSize
        content.imageProperties.reservedLayoutSize = imageSize
        ```
- label
    - 글씨가 잘리는 것을 막기 위해 numberOfLines를 0으로 설정하였습니다.
    - 첫화면 title label은 중앙 정렬을 해주었습니다.

3. Dynamic Type
- view에 사용자가 원하는 글씨사이즈로 볼수 있도록 하는 Dynamic Type을 적용하는 조건이 있어 코드로 구현을 할지 스토리보드로 나타내줄지에 대한 고민을 했었습니다.

    - 오토레이아웃이나 화면전환을 스토리보드로 구현하였다보니 통일성을 주기위해 Dynamic Type 또한 스토리보드에서 적용하였습니다.
    ![스크린샷 2023-11-08 오후 2.23.09(2).png](https://hackmd.io/_uploads/S1U5sqdmp.png)
    
     - 다만, 테이블 뷰의 List를 선택했을때 Detail 화면을 보여주는 경우 custoem cell로 구현을 하지않고, content configuration을 사용하여 원하는 image View를 설정하였다보니 Detail 화면의 Font의 경우엔 코드로 구현을 해줬습니다.
     ([Dynamic Type Size 참고](https://developer.apple.com/design/human-interface-guidelines/typography))
~~~swift
 content.textProperties.font = UIFont.preferredFont(forTextStyle: .title1)
        
content.secondaryTextProperties.font = UIFont.preferredFont(forTextStyle: .body)
~~~


4. accessibility 설정
[Writing Great Accessibility Labels (WWDC 2019)](https://developer.apple.com/videos/play/wwdc2019/254/)에 따르면
voice over시 유저가 들었을 때 이해할 수 있는 label을 사용하는 것이 중요합니다.
accessibility label의 경우 기본적으로 label의 내용이 설정됩니다.([accessibilitylabel 공식 문서](https://developer.apple.com/documentation/objectivec/nsobject/1615181-accessibilitylabel))
back button이 '메인 버튼', '한국의 출품작'처럼 이전 화면 title을 따르는데, 이를 들었을 때 조금 더 이해할 수 있는 label로 만들고자 하였습니다.
```swift
func setUpBackButtonAccessibilityLabel(to label: String) {
    let title = navigationItem.title
    let backButton = UIBarButtonItem(title: title, style: .plain, target: self, action: nil)

    navigationItem.backBarButtonItem = backButton
    navigationItem.backBarButtonItem?.accessibilityLabel = label
}
```
- `navigationItem.backBarButtonItem은 backButton이 있는 화면이 아닌 이전 화면에서 설정`해주어야 합니다.([backBarButtonItem 공식 문서](https://developer.apple.com/documentation/uikit/uinavigationitem/1624958-backbarbuttonitem)) 그래서 UIViewController을 extension하여 위 코드의 메서드를 구현했고, home viewcontroller와 list controller의 viewDidRoad에서 호출하고 있습니다.
- navigationItem.backBarButtonItem으로 UIKit에서 만든 back button에 접근 가능하다고 생각했으나 실행이 되지 않아 해결 방법으로 `새로운 back button을 만들어 설정`해주었습니다.


5. error발생시 alert 보여주도록 구현
- `json data를 decode할때 guard let에서 error를 던져주도록` 변경하였습니다.
- DecoderError enum을 만들어 assetNameError(DataAsset관련 에러)와 jsonDataError(decode관련 에러)를 case로 두고 있습니다.
```swift
struct AssetDecoder<Element: Decodable> {
    func parse(assetName: String) throws -> Element {
        guard let data = NSDataAsset(name: assetName)?.data else {
            throw DecoderError.assetNameError
        }
        
        let decoder = JSONDecoder()
        
        guard let jsonData = try? decoder.decode(Element.self, from: data) else {
            throw DecoderError.jsonDataError
        }
        
        return jsonData
    }
}
```
- ViewController에서 do-catch로 에러를 처리하도록 수정하였으며, `catch에서 alert를 띄어 유저가 error상황에 대해 인지`하도록 하였습니다.
- `UIViewController를 확장하여 error alert를 구현`하였습니다.
``` swift
func showAlert(_ error: Error) {
    let alertTitle: String = "오류"
    var message: String {
        switch error {
        case DecoderError.assetNameError:
            return "애셋네임을 알 수 없습니다."
        case DecoderError.jsonDataError:
            return "제이슨 데이터를 알 수 없습니다."
        default:
            return "시스템오류가 발생했습니다."
        }
    }
    let cancelTiltle: String = "취소"
    let retryTitle: String = "재시도"

    let alert = UIAlertController(title: alertTitle, message: message, preferredStyle: .alert)

    let cancelAction = UIAlertAction(title: cancelTiltle, style: .default)
    let retryAction = UIAlertAction(title: retryTitle, style: .default)

    alert.addAction(retryAction)
    alert.addAction(cancelAction)

    present(alert, animated: true)
}
```
- error에 따라 message가 변경되도록 연산프로퍼티로 구현하였습니다.
- `재시도, 취소 버튼`을 만들어 유저가 선택할 수 있도록 하였습니다.
    - 재시도 버튼 : 네트워크 상황시 데이터를 다시 가져오도록 하는 버튼이며, 현재 프로젝트에서는 정해진 Json data를 가져오는 것이기때문에 추가 실행을 구현하지 않고 일단 alert가 닫히도록만 구현하였습니다.
    - 취소 버튼 : 데이터가 제대로 decode되지 않은 상태로 남는 버튼입니다. 그래서 alert가 닫히도록만 구현하였습니다.

|asset error|json error|
|-|-|
|<img src="https://github.com/yagom-academy/ios-exposition-universelle/assets/126065608/9b136ed6-830f-4f25-bb54-b59d1cf03ee1" height="500">|<img src="https://github.com/yagom-academy/ios-exposition-universelle/assets/126065608/5269c8aa-44a6-401b-9eb0-a1ec582e3543" height="500">|




### 조언을 구하고 싶은 부분
1. Dynamic type을 Image에도 적용하는 것이 좋을까요?
현재 구현에서는 label만 dynamic type을 적용해주었는데 image view 또한 adjusts image size를 선택하면 변환이 되는 것을 확인하였습니다. 하지만 크기가 fix된 image에 경우에는 변환이 되지 않는데, 이 부분은 크기를 view의 frame 사이즈 기준으로 설정하는 등의 다른 방식으로 가능할 것 같습니다.

2. accessibility 적용 부분에서 조언을 구하고 싶습니다.
- 현재는 accessibility를 back bar button에만 수정해주고 있는데, 이 부분 외에 적용해주면 좋을 부분들이 있을까요? 그리고 Image에는 accessibility를 적용하고 있지 않은데, 어느 정도까지 적용을 해주면 좋을지 아직은 감이 잡히지 않습니다.
- back button에 accessibility를 적용하는 부분에서 navigationItem.backBarButtonItem으로 button을 가져올 수 있을 것으로 예상했으나 lldb 확인결과 nil로 확인되었습니다. 그래서 새로운 버튼을 넣어서 accessibility를 적용해주었는데, 이 방법이 아닌 UIKit이 만들어준 기본 backBarButton을 바로 사용해서 적용해볼 방법이 없을까요? 시도를 해보았으나 모두 실패하였습니다.

3. 고민되었던 점들이 어느정도 해결은 되었지만, 확신이 들지않아 전반적인 코드리뷰를 통해 조언을 얻고 싶습니다.

jyubong committed:
feat: ExpositionHomeViewController autoLayout 설정
스토리보드 stackView layout margins로 padding 20으로 설정
button 양 옆 image 크기 설정

Commit:
feat: TableView 리스트에 오토레이아웃 적용

jyubong committed:
feat: ExpositionItemDetailViewcontroller autolayout 설정
스토리보드로 imageView 고정
stackView padding 20으로 설정
scrollview superView에 top,bottom anchor 0으로 맞춤

Commit:
feat: 첫화면 세로로 볼수 있게 설정

jyubong committed:
feat: ExpositionHomeViewController dynamic font 적용
스토리보드로 dynamic 폰트 설정, automatically adjusts font 체크
subtitle Label 추가, Exposition에서 subtitle과 description 분리

Commit:
feat: 세번째 ViewController Dynamic Type 적용
두번쨰 ViewController 폰트 스타일 적용

jyubong committed:
refactor: appdelegate refactor, configureContent 메서드 추가

Commit:
feat: UIViewController+를 통해 알럿창 구현을 위한 메서드 구현

jyubong committed:
refactor: AssetDecoder 로직 수정, viewcontroller에 alert 띄우는 메서드 추가

Commit:
refactor: 주석 수정, 데이터네임 수정

jyubong committed:
feat: backbutton accessibilitylabel 설정 메서드 추가

delmaSong commented on Expo1900/Expo1900/Model/DecoderError.swift:
```diff
@@ -0,0 +1,13 @@
+enum DecoderError: Error {
+    case assetNameError
+    case jsonDataError
+}
```
error라는 이름이 매번 쓰이지 않더라도 error라고 알 수 있을 것 같아요

jyubong replied:
DecoderError에 error가 명시되어있어 각 case에서는 error을 없애주었습니다. -> assetName, jsonData

delmaSong commented on Expo1900/Expo1900/Controller/ExpositionItemListViewController.swift:
```diff
@@ -18,7 +18,16 @@ final class ExpositionItemListViewController: UIViewController {
+        } catch {
+            self.showAlert(error)
+        }
     }
```
에러 상황을 콘솔로 표시하는 게 아니라 사용자가 알 수 있도록 하는게 좋네요

delmaSong commented on Expo1900/Expo1900/Controller/UIViewController+.swift:
```diff
@@ -0,0 +1,44 @@
+            case DecoderError.jsonDataError:
+                return "제이슨 데이터를 알 수 없습니다."
+            default:
+                return "시스템오류가 발생했습니다."
```
사용자가 알게 하는 건 좋은데, 일반 사용자들은 애셋네임이 뭔지 제이슨 데이터가 뭔지 모를 것 같고 알 필요도 없을 것 같아요 😂

mireu79 replied:
피드백 주신대로 사용자 입장에선 애셋네임이 뭔지 제이슨 데이터가 뭔지 알 필요가 없을 거 같아 "데이터를 불러올 수 없습니다."라고 알려주고, 저희 입장에선 error가 무엇인지 구별할 수 있게 콘솔창에 print문으로 error를 표시할수 있게 해주었습니다.

delmaSong commented on Expo1900/Expo1900/Controller/UIViewController+.swift:
```diff
@@ -0,0 +1,44 @@
+            case DecoderError.assetNameError:
+                return "애셋네임을 알 수 없습니다."
+            case DecoderError.jsonDataError:
+                return "제이슨 데이터를 알 수 없습니다."
```
그냥 showAlert라고 하면 title, meesage등을 전달해서 설정해야하는 일반적인 alert느낌인데
여기는 에러 알럿을 보여주도록 되어있네요
동작을 잘 나타내도록 이름을 변경하는 편이 좋을 것 같아요

jyubong replied:
showAlert -> showErrorAlert 로 변경하여 error 관련 alert 임을 명시해주었습니다.

delmaSong reviewed (approved):
수고많으셨어요!
오토레이아웃 적용과 다이나믹 타입 적용이 잘 된 것 같군요 👍🏼

1. Dynamic type을 Image에도 적용하는 것이 좋을까요?
  - 보여지는 밸런스를 생각하면 같이 조절되는 편이 좋지 않을까 싶습니다. 텍스트 사이즈를 엄청나게 키운 경우에도 이미지는 커지지 않으니까 상대적으로 작아보이기도 하고, 크게 보고싶어서 조절하는 경우니 같이 조절되는 편이 좋을 것 같아요.
2. accessibility 적용 부분에서 조언을 구하고 싶습니다.
  - 사실 사용성을 생각하면 가능한 모든 부분에 접근성 지원을 하는 편이 좋지않을까 싶어요. 그렇지만 요구사항이나 개발 시간등의 여건등을 고려하여 적절한 범위를 정하는 편이 좋을 것 같습니다.
  - 이 부분은 저도 찾아보겠습니다~!

Commit:
refactor: Error처리 네이밍 및 Error 메서지  수정, 알럿창 함수 네이밍 수정

jyubong committed:
feat: image dynamic type 적용

extension UIImage로 새로운 사이즈 image 생성 메서드 추가
HomeViewcontroller와 DetailViewcontroller adjusts image size 적용

jyubong committed:
refactor: detail viewcontroller imageheight를 CGFloat으로 수정

jyubong commented:
image에도 dynamic type을 적용해주었습니다. 
image의 경우 크기 설정시 adjust image size를 클릭해도 dynamic이 적용이 되지 않았습니다.
그래서 세번째 뷰(detail view)의 이미지는 UIImage를 확장하여 원하는 높이를 설정하면 비율에 맞게 새로운 image를 생성하여 image view에 넣을 수 있도록 메서드를 추가하였습니다.

```swift
extension UIImage {
    func resized(height: CGFloat) -> UIImage {
        let ratio = self.size.width / self.size.height
        let newWidth = height * ratio
        let size = CGSize(width: newWidth, height: height)
        
        let renderer = UIGraphicsImageRenderer(size: size)
        let image = renderer.image { _ in
            draw(in: CGRect(origin: .zero, size: size))
        }
        
        return image
    }
}
```

|첫번째 화면|세번째화면|
|-|-|
|![](https://github.com/yagom-academy/ios-exposition-universelle/assets/126065608/f4de3b84-bc9c-4482-810a-033d0f0476a3)|![](https://github.com/yagom-academy/ios-exposition-universelle/assets/126065608/f8b01476-9f82-4e07-bd7a-255511e6c262)|

✔️ 두번째 tableview cell의 경우 cell의 넓이가 한정적입니다. 글씨가 커질시 높이가 길어지게 되는데 길이가 길 수록 table view 내의 cell들을 보기가 어렵다고 생각합니다. 그래서 image보다는 text를 키우는 것이 더 우선으로 생각되어 image에 dynamic type을 적용하지 않았습니다.

mireu79 committed:
Update README.md

jyubong committed:
Update README.md
